### ISSUE TITLE
Charm megatest hangs on gni-crayxc-smp builds

### ISSUE THREAD

**Issue by nitbhat** [UIUC-PPL/charm#2508]
Build Command: `./build charm++ gni-crayxc persistent smp --suffix=bug_prod_6.9 --with-production -j16`

Run Command: `srun -n 2 -c 16 ./pgm ++ppn 15 +commap 15,31 +pemap 0-14,16-30`

Output:
```Charm++> Running on Gemini (GNI) with 4 processes
Charm++> static SMSG
Charm++> SMSG memory: 19.8KB
Charm++> memory pool init block size: 8MB, total memory pool limit 0MB (0 means no limit)
Charm++> memory pool registered memory limit: 200000MB, send limit: 100000MB
Charm++> only comm thread send/recv messages
Charm++> Cray TLB page size: 8192K
Charm++> Running in SMP mode: 4 processes, 15 worker threads (PEs) + 1 comm threads per process, 60 PEs total
Charm++> The comm. thread both sends and receives messages
Converse/Charm++ Commit ID: v6.10.0-beta1-14-g918f46306
CharmLB> Load balancer assumes all CPUs are same.
Charm++> cpu affinity enabled.
Charm++> cpuaffinity PE-core map : 0-14,
Charm++> Running on 2 hosts (2 sockets x 16 cores x 2 PUs = 64-way SMP)
WARNING: Multiple PEs assigned to same core, recommend adjusting processor affinity or passing +CmiSleepOnIdle to reduce interference.
Megatest is running on 4 nodes 60 processors.
test 0: initiated [groupring (milind)]
test 0: completed (36.17 sec)
test 1: initiated [nodering (milind)]
test 1: completed (0.00 sec)
test 2: initiated [varsizetest (mjlang)]
test 2: completed (0.00 sec)
test 3: initiated [varsizetest2 (phil)]
test 3: completed (0.01 sec)
test 4: initiated [varraystest (milind)]
test 4: completed (0.00 sec)
test 5: initiated [groupcast (mjlang)]
test 5: completed (0.16 sec)
test 6: initiated [groupmulti (gengbin)]
test 6: completed (0.22 sec)
test 7: initiated [groupsectiontest (ebohm)]
test 7: completed (0.26 sec)
test 8: initiated [multisectiontest (ebohm)]
test 8: completed (0.84 sec)
test 9: initiated [nodecast (milind)]
test 9: completed (0.03 sec)
test 10: initiated [synctest (mjlang)]
test 10: completed (2.09 sec)
test 11: initiated [fib (jackie)]
test 11: completed (0.17 sec)
test 12: initiated [arrayring (fang)]
test 12: completed (1.62 sec)
test 13: initiated [packtest (fang)]
test 13: completed (0.19 sec)
test 14: initiated [queens (jackie)]
test 14: completed (0.35 sec)
test 15: initiated [migration (jackie)]
test 15: completed (0.62 sec)
test 16: initiated [marshall (olawlor)]
test 16: completed (0.40 sec)
test 17: initiated [priomsg (fang)]
test 17: completed (0.00 sec)
test 18: initiated [priotest (mlind)]
test 18: completed (0.00 sec)
test 19: initiated [rotest (milind)]
test 19: completed (0.03 sec)
test 20: initiated [statistics (olawlor)]
test 20: completed (0.00 sec)
test 21: initiated [templates (milind)]
test 21: completed (0.03 sec)
test 22: initiated [inherit (olawlor)]
test 22: completed (0.06 sec)
test 23: initiated [reduction (olawlor)]
test 23: completed (0.14 sec)
test 24: initiated [bitvector (jbooth)]
test 24: completed (0.00 sec)
test 25: initiated [immediatering (gengbin)]
test 25: completed (18.78 sec)
test 26: initiated [callback (olawlor)]
test 26: completed (0.41 sec)
test 27: initiated [inlineem (phil)]
test 27: completed (0.00 sec)
test 28: initiated [completion_test (phil)]
Starting test
Created detector, starting first detection
Started first test
Finished second test
Started third test
test 28: completed (0.56 sec)
test 29: initiated [groupdependence (nbhat4)]
test 29: completed (0.19 sec)
test 30: initiated [multi groupring (milind)]
test 30: completed (44.56 sec)
test 31: initiated [multi nodering (milind)]
test 31: completed (0.02 sec)
test 32: initiated [multi varsizetest (mjlang)]
test 32: completed (0.00 sec)
test 33: initiated [multi varsizetest2 (phil)]
test 33: completed (0.03 sec)
test 34: initiated [multi varraystest (milind)]
test 34: completed (0.00 sec)
test 35: initiated [multi groupcast (mjlang)]
test 35: completed (0.10 sec)
test 36: initiated [multi groupmulti (gengbin)]
test 36: completed (0.19 sec)
test 37: initiated [multi groupsectiontest (ebohm)]
test 37: completed (0.38 sec)
test 38: initiated [multi multisectiontest (ebohm)]
test 38: completed (0.87 sec)
test 39: initiated [multi nodecast (milind)]
test 39: completed (0.02 sec)
test 40: initiated [multi synctest (mjlang)]
test 40: completed (2.29 sec)
test 41: initiated [multi fib (jackie)]
test 41: completed (0.18 sec)
test 42: initiated [multi arrayring (fang)]
test 42: completed (3.21 sec)
test 43: initiated [multi packtest (fang)]
test 43: completed (0.45 sec)
test 44: initiated [multi migration (jackie)]
test 44: completed (0.82 sec)
test 45: initiated [multi marshall (olawlor)]
test 45: completed (0.11 sec)
test 46: initiated [multi priomsg (fang)]
test 46: completed (0.01 sec)
test 47: initiated [multi priotest (mlind)]
test 47: completed (0.02 sec)
test 48: initiated [multi statistics (olawlor)]
test 48: completed (0.00 sec)
test 49: initiated [multi reduction (olawlor)]
test 49: completed (0.63 sec)
test 50: initiated [multi immediatering (gengbin)]
test 50: completed (21.50 sec)
test 51: initiated [multi callback (olawlor)]
test 51: completed (0.36 sec)
test 52: initiated [all-at-once]
Starting test
Created detector, starting first detection
```

**Comment by nitbhat** [comment]
This issue also occurs for gni-crayxc-smp builds. 

On debugging, I found that this hang occurs during the second phase of completion detection in megatest. On starting first detection, the `start_detection` message which is broadcasted by one of the PEs doesn't reach all the PEs. I found that it reaches the level 1 handler (`_skipCldHandler`) and is enqueued in `schedQ`. However, it is never dequeued after that from the `schedQ` and hence the level 2 handler (`_processHandler`) is not executed on the message. I am still debugging to find out where the enqueued message is lost and why it is not dequeued.

**Comment by nitbhat** [comment]
I tried Sanjay's suggestion of making `start_detection` entry method `expedited`. However, that caused a hang (once in a while) at another place just after starting the first test. 

To reproduce bug:

1. Build `gni-crayxc-smp` or `gni-crayxc-persistent-smp` on Cori(both debug and production builds will work)

2. Allocate 1 node using salloc (`salloc --qos=interactive -C haswell --time=120 --nodes=1`)

3. Run a two-process job, each with 15 worker threads and 1 comm thread. 
`for i in {1..100}; do echo "Iteration $i***********************" && srun -n 2 -c 16 -N 1 ./pgm ++ppn 15 +pemap 0-14,16-30 +commap 15,31 ; done |& tee result`

**Comment by stwhite91** [comment]
I ran some tests on Cori today with a gni-craycx-smp build and found that this doesn't reproduce in a single process run. I confirmed it hangs for a 2 process run in the completion detection test (#28) of megatest sometimes, not all of the time. I also confirmed that marking everything in the completion detection library [expedited] doesn't fix the problem.

I think we are safe to release v6.10 with this bug unresolved since it is not a regression (v6.9.0 exhibits the same issue), and the issue has been narrowed down to multiprocess runs in SMP mode with completion detection, which no known applications rely on at this time.

**Comment by stwhite91** [comment]
Also, we don't yet know if this is confined to gni or not. @VenkatSNarayanan is going to do some runs on other systems to see if that's the case or not.

**Comment by ericjbohm** [comment]
I cleared the milestone from 6.10. Even if it isn't gni specific, its a
niche issue.

**Comment by matthiasdiener** [comment]
Megatest hangs on Blue Waters (gni-crayxe-smp) as well, in the `test 37: initiated [multi migration (jackie)]` test. Aprun command to reproduce:

``` bash
$ for i in $(seq 1 100); do aprun -n 2 -N 2 -d 8 ./pgm +ppn 15 +setcpuaffinity; done
```

**Comment by nitbhat** [comment]
> Megatest hangs on Blue Waters (gni-crayxe-smp) as well, in the `test 37: initiated [multi migration (jackie)]` test. Aprun command to reproduce:
> 
> ```shell
> $ for i in $(seq 1 100); do aprun -n 2 -N 2 -d 8 ./pgm +ppn 15 +setcpuaffinity; done
> ```

Oh okay, that's the original issue that was reported by Ted from Cray Inc. We weren't sure if the issue was resolved or not. But since you were able to reproduce the exact same issue, it looks like it wasn't resolved. Should we open a separate bug for that? Both bugs could be related where something weird happens to the message queue and the handler is not executed on some of the enqueued messages.

**Comment by matthiasdiener** [comment]
I think we could keep these issues in the same bug for now, as they seem to be related.

**Comment by nitbhat** [comment]
I further narrowed down the reproducibility test. The bug shows up in the "all-at-once" testcase when the following two tests are conducted - test id 7 (group section test) and test id 15 (migration test). 


This is the git diff from the current master that you'd need to reproduce the test.
```diff --git a/tests/charm++/megatest/megatest.C b/tests/charm++/megatest/megatest.C
index 2fb579b49..4b5bdb341 100644
--- a/tests/charm++/megatest/megatest.C
+++ b/tests/charm++/megatest/megatest.C
@@ -146,9 +146,14 @@ nextidx:
     CmiPrintf("test %d: initiated [all-at-once]\n", num);
     for (i=0; i<bank; i++) {
       if (!megatest_skip(tests[i].name)) {
+       //if(i==28||i==7||i==15) { - bug found
+       //if(i==28||i==15) { - bug not found
+       //if(i==28||i==7) { - bug not found
+       if(i==7||i==15) {
        acks_expected++;
        (tests[i].initiator)();
       }
+      }
     }
     megatest_finish();
     return;
@@ -187,10 +192,12 @@ main::main(CkArgMsg *msg)
   mainhandle = thishandle;
   if (nTests<=0)
     CkAbort("Megatest: No tests registered-- is MEGATEST_REGISTER_TEST malfunctioning?");
+/* All of tests 7,15 and 28 don't have any setup (moduleinit)
   for (i=0; i<nTests; i++)
     (tests[i].initializer)();
-  test_bank_size = nTests;
-  next_test_index = 0;
+ */
+ test_bank_size = nTests;
+  next_test_index = 60;
   next_test_number = 0;
   test_negate_skip=0;
   test_repeat = 0;```

**Comment by nitbhat** [comment]
I've made modifications to megatest to the simplest possible reproducible case in this branch: `nitin/gni-crayxc-smp-hang-debug`. I've added print statements in entry methods to help identify which EMs are executed and which are not.

**Comment by nitbhat** [comment]
This is fixed by https://github.com/UIUC-PPL/charm/pull/2551. Now that we have a fix for this, should we target it to 6.10?
